dhall/src/builtins.rs

```diff
@@ -401,6 +409,57 @@ fn apply_builtin(b: Builtin, args: Vec<Nir>, env: NzEnv) -> NirKind {
             }
             _ => Ret::DoneAsIs,
         },
+        (Builtin::TextReplace, [needle, replacement, haystack]) => {
+            match (&*needle.kind(), &*haystack.kind()) {
+                (TextLit(n_lit), TextLit(h_lit)) => {
+                    // The needle and the haystack need to be fully
+                    // evaluated as Text otherwise no progress can be made
+                    match (n_lit.as_text(), h_lit.as_text()) {
+                        (Some(n), Some(h)) => {
+                            // Helper to match a Nir as a text literal
+                            fn nir_to_string(n: &Nir) -> Option<String> {
+                                match &*n.kind() {
+                                    TextLit(r_lit) => r_lit.as_text(),
+                                    _ => None,
+                                }
+                            }
+
+                            // When the needle is empty the haystack is returned untouched
+                            if n.is_empty() {
+                                Ret::Nir(haystack.clone())
+                            // Fast case when replacement is fully evaluated
+                            } else if let Some(r) = nir_to_string(replacement) {
+                                Ret::Nir(Nir::from_text(h.replace(&n, &r)))
+                            } else {
+                                use std::iter::{once, repeat};
+
+                                let mut parts = h.split(&n).map(|s| {
+                                    InterpolatedTextContents::Text(
+                                        s.to_string(),
+                                    )
+                                });
+                                let replacement =
+                                    InterpolatedTextContents::Expr(
+                                        replacement.clone(),
+                                    );
+                                let first = parts.next();
+                                let rest = repeat(replacement)
+                                    .zip(parts)
+                                    .flat_map(|(r, p)| once(r).chain(once(p)));
```

Comment:
It looks like you could use https://docs.rs/itertools/0.9.0/itertools/trait.Itertools.html#method.interleave here maybe?

Reply:
Yes, I think this could work, I was looking for something like [`intercalate`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List.html#v:intercalate) (from Haskell), but now that I'm wrapping `replacement` as a single expr, `interleave` should be enough 👍

Reply:
`intersperse` works great here! :smile: